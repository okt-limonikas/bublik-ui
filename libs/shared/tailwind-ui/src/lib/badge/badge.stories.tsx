/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import React from 'react';
import { StoryFn, Meta } from '@storybook/react';

import { Badge, BadgeVariants } from './badge';

export default {
	component: Badge,
	title: 'components/Badge',
	decorators: [
		(Story) => (
			<div className="flex flex-wrap items-center gap-4 p-4 bg-white rounded-md">
				{Story()}
			</div>
		)
	]
} as Meta<typeof Badge>;

const SingleTemplate: StoryFn<typeof Badge> = (args) => (
	<Badge {...args}>Badge</Badge>
);

const MultipleTemplate: StoryFn<typeof Badge> = (args) => {
	return (
		<React.Fragment>
			<Badge {...args} variant={BadgeVariants.Primary}>
				Primary
			</Badge>
			<Badge {...args} variant={BadgeVariants.PrimaryActive}>
				Primary Active
			</Badge>
			<Badge {...args} variant={BadgeVariants.Expected}>
				Expected
			</Badge>
			<Badge {...args} variant={BadgeVariants.ExpectedActive}>
				Expected Active
			</Badge>
			<Badge {...args} variant={BadgeVariants.Unexpected}>
				Unexpected
			</Badge>
			<Badge {...args} variant={BadgeVariants.UnexpectedActive}>
				Unexpected Active
			</Badge>
			<Badge {...args} variant={BadgeVariants.Transparent}>
				Transparent
			</Badge>
		</React.Fragment>
	);
};

export const Primary = {
	render: SingleTemplate,
	args: { variant: BadgeVariants.Primary, isSelected: false }
};

export const All = {
	render: MultipleTemplate,
	args: { isSelected: false }
};

export const Custom = {
	render: SingleTemplate,
	args: {
		isSelected: false,
		className: 'bg-badge-16'
	}
};
