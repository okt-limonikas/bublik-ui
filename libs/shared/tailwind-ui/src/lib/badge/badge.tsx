/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ElementType, forwardRef } from 'react';

import { PolymorphicComponentPropWithRef, PolymorphicRef } from '../types';
import { cva, cn, VariantProps } from '../utils';

export const enum BadgeVariants {
	Primary = 'primary',
	Transparent = 'transparent',
	Expected = 'expected',
	Unexpected = 'unexpected',
	PrimaryActive = 'primary-active',
	ExpectedActive = 'expected-active',
	UnexpectedActive = 'unexpected-active'
}

export const badgeBaseStyles = cva({
	base: [
		'inline-flex',
		'items-center',
		'w-fit',
		'py-0.5',
		'px-2',
		'rounded',
		'border',
		'border-transparent',
		'leading-[1.125rem]',
		'text-[0.75rem]',
		'font-medium',
		'transition-colors'
	]
});

export const badgeVariantStyles = cva({
	variants: {
		variant: {
			[BadgeVariants.Transparent]: 'bg-transparent border-border-primary',
			[BadgeVariants.Primary]: 'bg-badge-0',
			[BadgeVariants.PrimaryActive]: 'bg-badge-9',
			[BadgeVariants.Expected]: 'text-text-expected bg-badge-3',
			[BadgeVariants.ExpectedActive]: 'bg-badge-3',
			[BadgeVariants.Unexpected]: 'bg-bg-fillError text-text-unexpected',
			[BadgeVariants.UnexpectedActive]: 'bg-bg-fillError'
		}
	}
});

export const badgeSelectedStyles = cva({
	variants: {
		variant: {
			[BadgeVariants.Transparent]: 'bg-primary-wash border-primary',
			[BadgeVariants.Primary]: 'bg-primary-wash border-primary',
			[BadgeVariants.Expected]: 'border-text-expected',
			[BadgeVariants.Unexpected]: 'border-text-unexpected',
			[BadgeVariants.PrimaryActive]: 'text-white bg-primary',
			[BadgeVariants.ExpectedActive]: 'text-white bg-bg-ok',
			[BadgeVariants.UnexpectedActive]: 'text-white bg-bg-error'
		}
	}
});

export type BadgeStylesProps = {
	variant?: VariantProps<typeof badgeVariantStyles>['variant'];
	className?: string;
	isSelected?: boolean;
	overflowWrap?: boolean;
};

const badgeStyles = (props: BadgeStylesProps) => {
	const { variant, isSelected, overflowWrap, className } = props;

	return cn(
		badgeBaseStyles(),
		badgeVariantStyles({ variant }),
		overflowWrap && 'overflow-wrap-anywhere',
		className,
		isSelected && badgeSelectedStyles({ variant })
	);
};

export type BadgeProps<E extends ElementType> = PolymorphicComponentPropWithRef<
	E,
	BadgeStylesProps
>;

export const Badge = forwardRef(
	<E extends ElementType>(
		{
			as,
			variant = BadgeVariants.Primary,
			isSelected,
			onClick,
			className,
			overflowWrap,
			children,
			...props
		}: BadgeProps<E>,
		ref?: PolymorphicRef<E>
	) => {
		const ComponentType = as ? as : onClick ? 'button' : 'div';

		return (
			<ComponentType
				className={badgeStyles({
					variant,
					isSelected,
					className,
					overflowWrap
				})}
				onClick={onClick}
				{...props}
				data-testid="tw-badge"
				ref={ref}
			>
				{children}
			</ComponentType>
		);
	}
);
