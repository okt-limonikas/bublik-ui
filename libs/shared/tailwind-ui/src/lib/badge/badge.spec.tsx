/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { it, describe, expect } from 'vitest';
import { render } from '@testing-library/react';

import { Badge } from './badge';

describe('components/Badge', () => {
	it('should match snapshot', () => {
		const { asFragment } = render(<Badge />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should display selected variant', () => {
		const { asFragment } = render(<Badge isSelected />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should display passed background color', () => {
		const { asFragment } = render(<Badge className="bg-badge-16" />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should display passed background color and override it when selected', () => {
		const { asFragment } = render(<Badge className="bg-badge-16" isSelected />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render button if onClick is passed', () => {
		const mock = vi.fn();

		const { getByRole } = render(<Badge onClick={mock}>Badge</Badge>);

		const badge = getByRole('button', { name: /Badge/i });

		expect(badge.tagName).toBe('BUTTON');
	});
});
