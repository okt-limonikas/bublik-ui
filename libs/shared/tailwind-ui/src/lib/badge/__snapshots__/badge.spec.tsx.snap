// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/Badge > should display passed background color 1`] = `
<DocumentFragment>
  <div
    class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors bg-badge-16"
    data-testid="tw-badge"
  />
</DocumentFragment>
`;

exports[`components/Badge > should display passed background color and override it when selected 1`] = `
<DocumentFragment>
  <div
    class="inline-flex items-center w-fit py-0.5 px-2 rounded border text-[0.75rem] font-medium transition-colors bg-primary-wash border-primary"
    data-testid="tw-badge"
  />
</DocumentFragment>
`;

exports[`components/Badge > should display selected variant 1`] = `
<DocumentFragment>
  <div
    class="inline-flex items-center w-fit py-0.5 px-2 rounded border text-[0.75rem] font-medium transition-colors bg-primary-wash border-primary"
    data-testid="tw-badge"
  />
</DocumentFragment>
`;

exports[`components/Badge > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors bg-badge-0"
    data-testid="tw-badge"
  />
</DocumentFragment>
`;
